pfPorts/ftpsesame/files/ftpsesame.8
.\"
.\" Copyright (c) 2004 Camiel Dobbelaar, <cd@sentia.nl>
.\"
.\" Permission to use, copy, modify, and distribute this software for any
.\" purpose with or without fee is hereby granted, provided that the above
.\" copyright notice and this permission notice appear in all copies.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
.\" WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
.\" ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
.\" WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
.\" ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
.\" OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"
.Dd January 2, 2004
.Dt FTPSESAME 8
.Os
.Sh NAME
.Nm ftpsesame
.Nd automagic packet filter configurator for FTP
.Sh SYNOPSIS
.Nm ftpsesame
.Op Fl d
.Op Fl D Ar level
.Op Fl i Ar interface
.Op Fl t Ar tag
.Op Fl q Ar queue
.Op Ar expression
.Sh DESCRIPTION
.Nm
tracks FTP control connections to be able to timely add
.Xr pf 4
packet filter rules that allow active and passive FTP data transfers
to commence.
This results in the efficiency of a packet filter, and the fine-grained
control of a proxy.
It can run on both FTP clients and FTP servers, or a firewall in front thereof.
.Pp
.Nm ftpsesame
uses
.Xr bpf 4
to get a copy of the data inside FTP control connections passing
the specified interface.
This data is searched for negotiations about data connections.
For passive mode, this is a port that a client should use to connect
to a server.
For active mode this is a port and an IP address that a server
should use to connect to a client.
.Pp
Assuming the FTP control connection is from $client to $server, and
$port is negotiated,
.Nm ftpsesame
adds one of the following rules to
.Xr pf 4 ,
inside an anchor named "ftpsesame".
.Pp
In case of active mode (PORT or EPRT):
.Bd -literal -offset 2n
pass in log quick inet proto tcp \e
    from $server to $client port $port flags S/SAFR keep state
.Ed
.Pp
In case of passive mode (PASV or EPSV):
.Bd -literal -offset 2n
pass in log quick inet proto tcp \e
    from $client to $server port $port flags S/SAFR keep state
.Ed
.Pp
In the special case of passive mode where the client is subject to
NAT, an extra rule is added, where $client_real is the client's
real IP address:
.Bd -literal -offset 2n
pass in log quick inet proto tcp \e
    from $client_real to $server port $port flags S/SAFR keep state
.Ed
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl D Ar level
Debug level, ranging from 0 to 7.
Higher is more verbose.
The default is 5.
(These levels correspond to the
.Xr syslog 3
levels.)
.It Fl d
Do not daemonize.
The process will stay in the foreground, logging to stderr.
.It Fl i Ar interface
Listen on
.Ar interface .
.It Fl t Ar tag
Create rules with tag
.Ar tag .
Also, option
.Ar quick
is not used.
This way the anchor always returns to the main ruleset, with the
tag set on approved FTP data connections.
The tag can then be used in pass rules below the anchor.
.It Fl q Ar queue
Create rules with queue
.Ar queue
appended.
.It Ar expression
Selects the packets that
.Xr bpf 4
will pass on to
.Nm ftpsesame.
An expression consist of the primitives described in
.Xr tcpdump 8 .
For efficiency, the expression should narrow down the traffic as
much as possible.
The default is "tcp and port 21".
.El
.Sh CONFIGURATION EXAMPLES
To use the rules set up by
.Nm ftpsesame ,
.Xr pf.conf 5
should be modified to use the anchor.
Below are examples that show most of the possibilities.
.Bd -literal -offset 2n
# (1) Allow internal LAN to access FTP servers anywhere.
#     (passive mode only if $lan is subject to NAT, see below)
# cmd: ftpsesame -i $ext_if
anchor "ftpsesame/*" on { $int_if, $ext_if }
pass in  quick on $int_if proto tcp from $lan to any port 21 keep state
pass out quick on $ext_if proto tcp from $lan to any port 21 keep state

# (2) Allow access to FTP servers in the DMZ, while queueing the data
#     transfers.
# cmd: ftpsesame -i $dmz_if -q ftp
queue ftp bandwidth 10%
# passive
anchor "ftpsesame/*" in  on $ext_if proto tcp from any to { $ftp_servers }
anchor "ftpsesame/*" out on $dmz_if proto tcp from any to { $ftp_servers }
# active
anchor "ftpsesame/*" in  on $dmz_if proto tcp from { $ftp_servers } to any
anchor "ftpsesame/*" out on $ext_if proto tcp from { $ftp_servers } to any
# ... rules for port 21 omitted ...

# (3) Allow access to FTP servers in the DMZ, using tagging for fine
#     grained control.
# cmd: ftpsesame -i $dmz_if -t ftpok
anchor "ftpsesame/*"
# passive
pass in  quick on $ext_if proto tcp from to any to $realftp \e
    port > 49151 tagged ftpok keep state
pass in  quick on $ext_if proto tcp from any to $msftp \e
    port 1023 >< 5001 tagged ftpok keep state
pass out quick on $dmz_if all tagged ftpok keep state
# active
pass in  quick on $dmz_if proto tcp from $realftp to any \e
    tagged ftpok keep state
pass in  quick on $dmz_if proto tcp from $msftp port 20 to any \e
    tagged ftpok keep state
pass out quick on $ext_if all tagged ftpok keep state
# ... rules for port 21 omitted ...
.Ed
.Pp
Keep state is mandatory on the control connection (port 21), because
.Nm ftpsesame
checks that these connections are in the statetable, before allowing
data connections.
.Sh NAT ISSUES
To enable passive mode connections from clients subject to NAT,
.Nm ftpsesame
must listen on the interface where the NAT has already taken place,
typically the external interface.
Otherwise
.Nm ftpsesame
cannot add the extra rule with the real client address.
Note that this does not have to be a problem, if there are other
rules taking care of that.
.Pp
Active mode connections from clients subject to NAT are not supported,
because it requires commands inside the control connection to be
rewritten.
.Xr ftp-proxy 8
can be used for that purpose.
.Sh SECURITY
Ports below 1024 are not allowed.
.Pp
The negotiated IP address for active mode is ignored for security
reasons.
This makes third party file transfers impossible.
.Pp
.Nm ftpsesame
checks that the ftp control connection is in the packet filter
statetable before it will trust its contents.
.Pp
.Nm ftpsesame
chroots to "/var/empty" and changes to user "proxy" to drop privileges.
It does keep a file descriptor to both
.Xr bpf 4
and
.Xr pf 4
so it is still very powerful.
.Pp
The concept of rebuilding a TCP session from IP packets on a passive
listener is fundamentally flawed.
.Nm ftpsesame
would need a full TCP stack that emulates each endpoint perfectly
to be safe from all evasion techniques.
.Xr ftp-proxy 8
does not have this problem.
Ofcourse
.Nm ftpsesame
tries hard to detect irregularities.
The following are not allowed: IP options, IP fragments, ttl below
2, lines without end-of-line marker.
Also,
.Nm ftpsesame
only looks at the first line of whatever the server has to say in
response to a client.
.Pp
It is recommended to run
.Nm ftpsesame
on an interface where packets from untrusted hosts are already
scrubbed.
Also, anchors and tags can be used to restrict the allowed addresses
and portranges.
.Sh CAVEATS
There is always a slight delay before a rule gets added to the
anchor.
If the data connection is opened within this period, it will get blocked.
If
.Nm ftpsesame
runs on a firewall between the client and server, this only introduces
a slight delay, as TCP will retry within seconds.
If
.Nm ftpsesame
runs on a client or server itself, the network stack reacts differently
on blocked packets ("no route to host") and either active or passive
may mode not work.
.Sh SEE ALSO
.Xr bpf 4 ,
.Xr ftp 1 ,
.Xr ftp-proxy 8 , 
.Xr pf 4 ,
.Xr pf.conf 5 ,
.Xr tcpdump 8


pfPorts/ftpsesame/files/ftpsesame.c
/*
 * Copyright (c) 2004 Camiel Dobbelaar, <cd@sentia.nl>
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

#include <sys/types.h>
#include <sys/socket.h>
#include <sys/ioctl.h>

#include <net/if.h>
#include <net/pfvar.h>
#include <net/ppp_defs.h>
#include <netinet/if_ether.h>
#include <netinet/in.h>
#include <netinet/in_systm.h>
#include <netinet/ip.h>
#include <netinet/tcp.h>
#include <netinet/tcp_fsm.h>
#include <arpa/inet.h>

#include <ctype.h>
#include <err.h>
#include <errno.h>
#include <pcap.h>
#include <pwd.h>
#include <stdarg.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <syslog.h>
#include <time.h>
#include <unistd.h>

#include "filter.h"
#include "state.h"

#define NOPRIV_USER	"proxy"
#define CHROOT_DIR	"/var/empty"

#define PCAP_TO_MS	500
#define PURGE_INTERVAL	60
#define SNAPLEN		500

#define IPSTR_LEN	16
#define	IP_HDRLEN	(sizeof(struct ip))
#define TH_MASK		(TH_FIN|TH_SYN|TH_RST|TH_ACK)

int	 clienttalk(struct state *, char *);
void	 servertalk(struct state *, char *);
int	 parse_port(char *, int);
struct state *tcp_state(struct ip *, struct tcphdr *, int *);
void	 process_pkt(u_char *, struct pcap_pkthdr *, u_char *);
char	*copy_argv(char * const *);
int	 drop_privs(void);
void	 logmsg(int, const char *, ...);
void	 sig_alarm(int);
void	 sig_close(int);
void	 usage(void);

enum { PARSE_PORT, PARSE_PASV, PARSE_EPSV, PARSE_EPRT };
enum { ACTIVE, PASSIVE };

u_int32_t ps_proc;
int daemonize, linktype, loglevel;
unsigned int linklen;
volatile sig_atomic_t gotsig_alarm, gotsig_close;
extern char *__progname;

int
clienttalk(struct state *s, char *cbuf)
{
	size_t buflen;

	/* Cbuf length is ultimately bound by the snaplen. */

	logmsg(LOG_DEBUG, "#%d client: %s", s->id, cbuf);

	buflen = strlen(cbuf) + 1;
	if (buflen > s->clientbuflen) {
		if (s->clientbuf != NULL)
			free(s->clientbuf);
		/* What reduces the number of mallocs? */
		buflen += 42;
		s->clientbuf = malloc(buflen);
		if (s->clientbuf == NULL) {
			logmsg(LOG_ERR, "clienttalk malloc");
			s->clientbuflen = 0;
			return (0);
		}
		s->clientbuflen = buflen;
	}

	strlcpy(s->clientbuf, cbuf, s->clientbuflen);

	return (1);
}

void
servertalk(struct state *s, char *sbuf)
{
	struct in_addr *src, *dst, *realsrc;
	char *cbuf, ipstr[IPSTR_LEN];
	int mode, port;

	cbuf = s->clientbuf;

	logmsg(LOG_DEBUG, "#%d server: %s", s->id, sbuf);

	/* Passive mode (PASV). */
	if (strcasecmp("PASV", cbuf) == 0 &&
	    strncasecmp("227 ", sbuf, strlen("227 ")) == 0) {
		logmsg(LOG_INFO, "#%d passive: %s", s->id, sbuf);
		mode = PASSIVE;
		src = &s->ip_src;
		dst = &s->ip_dst;
		port = parse_port(sbuf, PARSE_PASV);
	}
	/* Active mode (PORT). */
	else if (strncasecmp("PORT ", cbuf, strlen("PORT ")) == 0 &&
	    strncasecmp("200 ", sbuf, strlen("200 ")) == 0) {
		logmsg(LOG_INFO, "#%d active: %s", s->id, cbuf);
		mode = ACTIVE;
		src = &s->ip_dst;
		dst = &s->ip_src;
		port = parse_port(cbuf, PARSE_PORT);
	}
	/* Extended passive mode (EPSV). */
	else if (strcasecmp("EPSV", cbuf) == 0 &&
	    strncasecmp("229 ", sbuf, strlen("229 ")) == 0) {
		logmsg(LOG_INFO, "#%d ext. passive: %s", s->id, sbuf);
		mode = PASSIVE;
		src = &s->ip_src;
		dst = &s->ip_dst;
		port = parse_port(sbuf, PARSE_EPSV);
	}
	/* Extended active mode (EPRT). */
	else if (strncasecmp("EPRT ", cbuf, strlen("EPRT ")) == 0 &&
	    strncasecmp("200 ", sbuf, strlen("200 ")) == 0) {
		logmsg(LOG_INFO, "#%d ext. active: %s", s->id, cbuf);
		mode = ACTIVE;
		src = &s->ip_dst;
		dst = &s->ip_src;
		port = parse_port(cbuf, PARSE_EPRT);
	}	
	/* No match. */
	else
		return;

	if (port == -1) {
		logmsg(LOG_WARNING, "#%d parse_port failed: client '%s', " \
		    "server '%s'", s->id, cbuf, sbuf);
		return;
	}

	if (port < 1024) {
		logmsg(LOG_WARNING, "#%d denying low port %d request", s->id,
		    port);
		return;
	}

	strlcpy(ipstr, inet_ntoa(*dst), sizeof ipstr);
	logmsg(LOG_INFO, "#%d allowing %s to %s port %d", s->id,
	    inet_ntoa(*src), ipstr, port);

	realsrc = NULL;
	if (mode == PASSIVE &&
	    memcmp(&s->ip_src, &s->ip_src_real, sizeof s->ip_src) != 0) {
		realsrc = &s->ip_src_real;
		logmsg(LOG_INFO, "#%d passive/nat: also allowing %s to " \
		    "%s port %d", s->id, inet_ntoa(*realsrc), ipstr, port);
	}

	if (filter_allow(s->id, src, realsrc, dst, port))
		s->rule_ts = time(NULL);
	else
		logmsg(LOG_ERR, "#%d filter_allow failed: %s", s->id,
		    strerror(errno));
}

int
parse_port(char *buf, int mode)
{
	unsigned int	i, port, v[6];
	char		c1, c2, *p;

	/* Find the last space or left-parenthesis. */
	for (p = buf + strlen(buf); p > buf; p--)
		if (*p == ' ' || *p == '(')
			break;
	if (p == buf)
		return (-1);

	/* The %c format specifier is used to detect trailing characters. */
	switch (mode) {
	case PARSE_PORT:
		i = sscanf(p, " %u,%u,%u,%u,%u,%u%c", &v[0], &v[1], &v[2],
		    &v[3], &v[4], &v[5], &c1);
		if (i != 6 || v[0] > 255 || v[1] > 255 || v[2] > 255 ||
		    v[3] > 255 || v[4] > 255 || v[5] > 255)
			return (-1);
		port = (v[4] << 8) | v[5];
		break;
	case PARSE_PASV:
		i = sscanf(p, "(%u,%u,%u,%u,%u,%u)%c%c", &v[0], &v[1], &v[2],
		    &v[3], &v[4], &v[5], &c1, &c2);
		if (i != 6) {
			/* Microsoft FTP server prints a trailing dot. */
			if (i != 7 || c1 != '.')
				return (-1);
		}
		if (v[0] > 255 || v[1] > 255 || v[2] > 255 ||
		    v[3] > 255 || v[4] > 255 || v[5] > 255)
			return (-1);
		port = (v[4] << 8) | v[5];
		break;
	case PARSE_EPSV:
		i = sscanf(p, "(|||%u|)%c", &port, &c1);
		if (i != 1 || port > 65535)
			return (-1);
		break;
	case PARSE_EPRT:
		/* Our EPRT support is limited to IPv4. */
		i = sscanf(p, " |1|%u.%u.%u.%u|%u|%c", &v[0], &v[1], &v[2],
		    &v[3], &port, &c1);
		if (i != 5 || v[0] > 255 || v[1] > 255 || v[2] > 255 ||
		    v[3] > 255 || port > 65535)
			return (-1);
		break;
	default:
		return (-1);
	}

	return (port);
}

struct state *
tcp_state(struct ip *ip, struct tcphdr *tcp, int *isclient)
{
	struct state *s;

	s = state_find(ip, tcp, isclient);

	if (s == NULL) {
		if ((tcp->th_flags & TH_MASK) == TH_SYN) {
			char ipstr[IPSTR_LEN];

			s = state_new(ip, tcp);
			if (s == NULL) {
				logmsg(LOG_ERR, "cannot add session");
				return (NULL);
			}
			strlcpy(ipstr, inet_ntoa(ip->ip_src), sizeof ipstr);
			logmsg(LOG_INFO, "#%d session init: client %s:%d, " \
			    "server %s:%d", s->id, ipstr, ntohs(tcp->th_sport),
			    inet_ntoa(ip->ip_dst), ntohs(tcp->th_dport));

			s->tcps = TCPS_SYN_SENT;
		}
		/* SYN has no payload. */
		return (NULL);
	}
			
	if (tcp->th_flags & (TH_FIN|TH_RST)) {
		logmsg(LOG_INFO, "#%d session finish", s->id);
		state_remove(s);
		return (NULL);
	}

	switch (s->tcps) {
	case TCPS_SYN_SENT:
		/* Expect a SYN/ACK from the server. */
		if (*isclient || (tcp->th_flags & TH_MASK) != (TH_SYN|TH_ACK))
			break;
		/*
		 * Now is a good time to check that pf allowed this connection.
		 * (pf does not filter for us, as we are on bpf)
		 */
		if (filter_lookup(&s->ip_src, &s->ip_dst, s->s_port, s->d_port,
		    &s->ip_src_real))
			s->tcps = TCPS_SYN_RECEIVED;
		else
			logmsg(LOG_WARNING, "#%d session not in statetable",
			    s->id);
		/* SYN/ACK has no payload. */
		break;
	case TCPS_SYN_RECEIVED:
		/* Expect an ACK from the client to complete the handshake. */
		if (*isclient && (tcp->th_flags & TH_MASK) == TH_ACK) {
			s->tcps = TCPS_ESTABLISHED;
			return (s);
		}
		break;
	case TCPS_ESTABLISHED:
		if ((tcp->th_flags & TH_MASK) == TH_ACK)
			return (s);
		break;
	default:
		logmsg(LOG_ERR, "#%d unknown tcp state: %d", s->id, s->tcps);
	}

	return (NULL);
}

void
process_pkt(u_char *notused, struct pcap_pkthdr *h, u_char *p)
{
	struct ip *ip;
	struct tcphdr *tcp;
	struct state *s;
	int isclient;
	unsigned int i, ip_len, tcp_datalen, tcp_off;
	char *tcp_data;

	ps_proc++;

	if (h->caplen != h->len) {
		logmsg(LOG_WARNING, "drop: short capture");
		return;
	}

	/*
	 * We don't want to grok all linktypes to be able to check if the
	 * packet is of type IP, so we just assume it is.  The user must
	 * supply a seriously bad expression on the commandline anyway for
	 * non-IP packets to get here.  In that case, the checks below should
	 * suffice, but they'll get a bit more logging.
	 */

	if (h->len < linklen + IP_HDRLEN) {
		logmsg(LOG_WARNING, "drop: short ip");
		return;
	}

	ip = (struct ip *)(p + linklen);

	if (ip->ip_v != 4) {
		logmsg(LOG_WARNING, "drop: not ipv4");
		return;
	}
	if ((ip->ip_hl << 2) != IP_HDRLEN) {
		logmsg(LOG_WARNING, "drop: ip options");
		return;
	}

	ip_len = ntohs(ip->ip_len);

	if (h->len < linklen + ip_len) {
		logmsg(LOG_WARNING, "drop: long ip");
		return;
	}
	if ((ntohs(ip->ip_off) & (IP_MF | IP_OFFMASK)) != 0) {
		logmsg(LOG_WARNING, "drop: fragmented");
		return;
	}
	if (ip->ip_ttl < 2) {
		logmsg(LOG_WARNING, "drop: ttl too low");
		return;
	}
	if (ip->ip_p != IPPROTO_TCP) {
		logmsg(LOG_WARNING, "drop: not tcp");
		return;
	}
	if (ip_len < IP_HDRLEN + sizeof(struct tcphdr)) {
		logmsg(LOG_WARNING, "drop: tcp short");
		return;
	}

	tcp = (struct tcphdr *)(p + linklen + IP_HDRLEN);

	/* Run our little state machine. */
	s = tcp_state(ip, tcp, &isclient);
	if (s == NULL)
		return;

	tcp_off = tcp->th_off << 2;
	if (ip_len < IP_HDRLEN + tcp_off) {
		logmsg(LOG_WARNING, "drop: tcp_off off");
		return;
	}
	tcp_datalen = ip_len - (IP_HDRLEN + tcp_off);
	if (tcp_datalen == 0)
		return;
	tcp_data = ((char *)tcp) + tcp_off;

	/*
	 * We are going to be strict here and only look at properly
	 * terminated lines.  The lines we are looking for are small
	 * enough to always fit in one IP packet.  If not, tough luck.
	 *
	 * If there are multiple lines in one packet, we only look at
	 * the first one.
	 */

	for (i = 0; i < tcp_datalen; i++)
		if (tcp_data[i] == '\r' || tcp_data[i] == '\n') {
			tcp_data[i] = '\0';
			break;
		}
	if (i == tcp_datalen) {
		if (s->clientbuf != NULL)
			s->clientbuf[0] = '\0';
		return;
	}

	if (isclient) {
		if (!clienttalk(s, tcp_data))
			state_remove(s);
		return;
	}

	if (s->clientbuf == NULL || s->clientbuf[0] == '\0')
		return;

	servertalk(s, tcp_data);
	s->clientbuf[0] = '\0';
}

int
main(int argc, char **argv)
{
	struct bpf_insn blockall[] = { BPF_STMT(BPF_RET+BPF_K, 0) };
	struct bpf_program bprog, wfilter;
	struct pcap_stat pstat;
	char errbuf[PCAP_ERRBUF_SIZE];
	char *filter, *interface, *qname, *tagname;
	pcap_t *hpcap;
	int ch, immediate;

	/* Defaults. */
	daemonize = 1;
	loglevel = LOG_NOTICE;
	interface = NULL;
	tagname = NULL;
	qname = NULL;
	filter = "tcp and port 21";

	while ((ch = getopt(argc, argv, "D:di:q:t:")) != -1) {
		switch (ch) {
		case 'D':
			loglevel = atoi(optarg);
			if (loglevel < LOG_EMERG || loglevel > LOG_DEBUG)
				usage();
			break;
		case 'd':
			daemonize = 0;
			break;
		case 'i':
			interface = optarg;
			break;
		case 't':
			if (strlen(optarg) >= PF_TAG_NAME_SIZE)
				errx(1, "tag too long");
			tagname = optarg;
			break;
		case 'q':
			if (strlen(optarg) >= PF_QNAME_SIZE)
				errx(1, "queue too long");
			qname = optarg;
			break;
		default:
			usage();
		}
	}
	argc -= optind;
	argv += optind;

	if (argc) {
		filter = copy_argv(argv);
		if (filter == NULL)
			err(1, "copy_argv");
	}

	/* Check for root to save the user from cryptic failure messages. */
	if (getuid() != 0)
		errx(1, "needs to start as root");

	filter_init(qname, tagname);

	if (interface == NULL) {
		interface = pcap_lookupdev(errbuf);
		if (interface == NULL)
			errx(1, "%s", errbuf);
	}
	hpcap = pcap_open_live(interface, SNAPLEN, 0, PCAP_TO_MS, errbuf);
	if (hpcap == NULL)
		errx(1, "%s", errbuf);

	linktype = pcap_datalink(hpcap);
	switch (linktype) {
	case DLT_EN10MB:
		linklen = sizeof(struct ether_header);
		break;
	case DLT_PPP:
		linklen = PPP_HDRLEN;
	case DLT_LOOP:
		linklen = 4;
		break;
	default:
		errx(1, "%s: unsupported interface type", interface);
	}

	if (pcap_compile(hpcap, &bprog, filter, 1, 0) < 0 ||
	    pcap_setfilter(hpcap, &bprog) < 0)
		errx(1, "filter '%s': %s", filter, pcap_geterr(hpcap));

	immediate = 1;
	if (ioctl(pcap_fileno(hpcap), BIOCIMMEDIATE, &immediate) < 0)
		err(1, "BIOCIMMEDIATE");
	wfilter.bf_len = 1;
	wfilter.bf_insns = blockall;
	if (ioctl(pcap_fileno(hpcap), BIOCSETWF, &wfilter) < 0)
		err(1, "BIOCSETWF");
	if (ioctl(pcap_fileno(hpcap), BIOCLOCK) < 0)
		err(1, "BIOCLOCK");

	if (daemonize) {
		if (daemon(0, 0) == -1)
			err(1, "daemon");
		openlog(__progname, LOG_PID | LOG_NDELAY, LOG_DAEMON);
	}

	/* Use logmsg for output from here on. */
	
	if (!drop_privs()) {
		logmsg(LOG_ERR, "drop_privs: %s", strerror(errno));
		exit(1);
	}

	gotsig_alarm = gotsig_close = 0;
	signal(SIGALRM, sig_alarm);
	signal(SIGINT, sig_close);
	signal(SIGQUIT, sig_close);
	signal(SIGTERM, sig_close);
	alarm(PURGE_INTERVAL);

	state_init();
	ps_proc = 0;

	for (;;) {
		/* Step up... Step up... Step in the Arena. */
		pcap_dispatch(hpcap, 0, (pcap_handler)process_pkt, NULL);

		if (gotsig_close)
			break;

		if (gotsig_alarm) {
			logmsg(LOG_DEBUG, "sessions after purging: %d",
				state_purge(0));
			gotsig_alarm = 0;
			alarm(PURGE_INTERVAL);
		}
	}
	
	state_purge(1);

	pcap_close(hpcap);

	if (daemonize)
		closelog();

	return (0);
}

char *
copy_argv(char * const *argv)
{
	size_t len = 0, n;
	char *buf;

	if (argv == NULL)
		return (NULL);

	for (n = 0; argv[n]; n++)
		len += strlen(argv[n]) + 1;
	if (len == 0)
		return (NULL);

	buf = malloc(len);
	if (buf == NULL)
		return (NULL);

	strlcpy(buf, argv[0], len);
	for (n = 1; argv[n]; n++) {
		strlcat(buf, " ", len);
		strlcat(buf, argv[n], len);
	}
	return (buf);
}

int
drop_privs(void)
{
	struct passwd *pw;
	gid_t gidset[1];

	pw = getpwnam(NOPRIV_USER);
	if (pw == NULL)
		return (0);

	tzset();

	if (chroot(CHROOT_DIR) != 0)
		return (0);
	chdir("/");

	gidset[0] = pw->pw_gid;
	if (setgroups(1, gidset) == -1 || setegid(pw->pw_gid) == -1 ||
	    setgid(pw->pw_gid) == -1 || seteuid(pw->pw_uid) == -1 ||
	    setuid(pw->pw_uid) == -1)
		return (0);
	
	return (1);
}

void
logmsg(int pri, const char *message, ...)
{
	va_list ap;
	va_start(ap, message);

	if (pri > loglevel)
		return;
		
	if (!daemonize) {
		vfprintf(stderr, message, ap);
		fprintf(stderr, "\n");
	} else
		vsyslog(pri, message, ap);
	va_end(ap);
}

void
sig_alarm(int sig)
{
	gotsig_alarm = 1;
}

void
sig_close(int sig)
{
	gotsig_close = 1;
}

void
usage(void)
{
	fprintf(stderr, "usage: %s [-d] [-D level] [-i interface] " \
	    "[-q queue] [-t tag] [expression]\n", __progname);
	exit(1);
}
